app/services/CartServiceImpl.java

```diff
@@ -0,0 +1,156 @@
+import io.sphere.sdk.customobjects.commands.CustomObjectUpsertCommand;
+import io.sphere.sdk.customobjects.queries.CustomObjectByKeyGet;
+import io.sphere.sdk.models.Address;
+import io.sphere.sdk.models.AddressBuilder;
+import io.sphere.sdk.models.DefaultCurrencyUnits;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.products.attributes.AttributeAccess;
+import pactas.models.PactasContract;
+import pactas.models.PactasCustomer;
+import play.Logger;
+import play.mvc.Http;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.springframework.util.Assert.notNull;
+
+public class CartServiceImpl extends AbstractShopService implements CartService {
+
+    public final static String FREQUENCY = "cart-frequency";
+    public final static String ID_MONTHLY = "pactas4";
+    public final static String ID_TWO_WEEKS = "pactas2";
+    public final static String ID_WEEKLY = "pactas1";
+
+    public CartServiceImpl(final SphereClient sphereClient) {
+        super(sphereClient);
+    }
+
+    @Override
+    public Cart getOrCreateCart(final Http.Session session) {
+        notNull(session, "Session is null, unable to create or get Cart");
+        return Optional.ofNullable(session.get(SessionKeys.CART_ID))
+                .map(String::valueOf)
+                .map(cardId -> {
+                            final Cart cart = sphereClient().execute(CartByIdGet.of(cardId)).toCompletableFuture().join();
+                            Logger.debug("Fetched existing Cart[cartId={}]", cart.getId());
+                            return cart;
+                        }
+                )
+                .orElseGet(() -> {
+                    final Cart cart = sphereClient().execute(CartCreateCommand.of(CartDraft.of(DefaultCurrencyUnits.EUR))).toCompletableFuture().join();
+                    Logger.debug("Created new Cart[cartId={}]", cart.getId());
+                    session.put(SessionKeys.CART_ID, cart.getId());
+                    return cart;
+                });
+    }
+
+    @Override
+    public Cart clearCart(final Cart cart) {
+        final List<RemoveLineItem> items = cart.getLineItems().stream().map((item) -> {
+            final RemoveLineItem removeLineItem = RemoveLineItem.of(item, 1);
+            return removeLineItem;
+        }).collect(Collectors.toList());
+        final Cart result = sphereClient().execute(CartUpdateCommand.of(cart, items)).toCompletableFuture().join();
+        clearFrequency(result.getId());
+        return result;
+    }
+
+    private void clearFrequency(final String cartId) {
+        Logger.debug("Clearing frequency");
+        final Optional<CustomObject<JsonNode>> result = Optional.ofNullable(
+                sphereClient().execute(CustomObjectByKeyGet.of(FREQUENCY, cartId)).toCompletableFuture().join());
+        if (result.isPresent()) {
```

Comment:
We try to avoid the `isPresent` calls because you tend to easier call `Object.get`.

```
result.ifPresent(customObject -> {
    Logger.debug("Fetched existing CustomObject: {}", customObject);//here is the object, not the optional
    final CustomObject<JsonNode> cleared =  sphereClient().execute(CustomObjectDeleteCommand.of(FREQUENCY, cartId)).toCompletableFuture().join();
    Logger.debug("Cleared CustomObject: {}", cleared);
});
```